src/lib/sample_sheet.rs

```diff
@@ -0,0 +1,587 @@
+        let mut line_index = start_line_index;
+        if records.len() == line_index {
+            // no "[Demux]" found
+            return Err(SampleSheetError::NoDemuxHeader);
+        }
+        line_index += 1; // skip over "[Demux]"
+        let mut argv: Vec<String> = vec![TOOL_NAME.to_string()];
+        while line_index < records.len() {
+            let record = &records[line_index];
+            if !record.is_empty() && &record[0] == "[Data]" {
+                break;
+            }
+            if record.len() >= 2 {
+                // Command line arguments and their values are not checked here to be valid.  This
+                // happens below when we try to build the opts.
+                let key = record[0].to_string();
+                let argument = format!("--{}", key);
+                argv.push(argument);
+
+                // For boolean options, only the presence of a key and the value is ignored.  Thus,
+                // when an empty value is given, assume the argument is a boolean.
+                if !record[1].is_empty() {
+                    argv.push(record[1].to_string());
+                }
+            }
+            line_index += 1;
+        }
+
+        // Build the opts given the arguments.  This is where any unknown command line argument
+        // will raise an error, and illegal value too.
+        match opts.try_update_from(argv.into_iter()) {
+            Ok(()) => Ok(line_index),
+            Err(err) => {
+                let kind = err.kind.as_str().unwrap_or("Unknown error").to_string();
+                let mut args: Vec<String> = vec![];
+                for string in err.info {
+                    args.push(string.split(' ').next().unwrap_or(&string).to_string());
+                }
+                Err(SampleSheetError::DemuxOptionsParsing { kind, args: args.join(", ") })
+            }
+        }
+    }
+
+    /// Reads from a Sample Sheet represented as a vector if `StringRecord`s.  Skips over lines
+    /// optionally until the `[Demux]` section is found, upating the given command line options,
+    /// and then the `[Data]` section, to parse a list of samples.
+    fn from_sample_sheet_string_records(
+        records: &[StringRecord],
+        opts: &Opts,
+    ) -> Result<Self, SampleSheetError> {
+        if records.is_empty() {
+            return Err(SampleSheetError::Empty);
+        }
+
+        let mut line_index = 0;
+
+        // ignore data pre-"[Demux]"
+        while line_index < records.len()
+            && !records[line_index].is_empty()
+            && &records[line_index][0] != "[Demux]"
+        {
+            line_index += 1;
+        }
+
+        // Parse demultiplexing settings
+        let mut opts = opts.clone();
+        let mut line_index = SampleSheet::slurp_demux_opts(records, line_index, &mut opts)?;
+        if records.len() == line_index {
+            // no "[Data]" found
+            return Err(SampleSheetError::NoData);
+        }
```

Comment:
I think I would refactor this to a function that takes the records and returns a `HashMap<String,Vec<StringRecord>>`, with the key being the section of the sample sheet, and the value being the records from that section.  I feel like relying on detection of `[Data]` could bite us, rather than allowing any section to come next (including yet to be invented sections).